backend-docker/pier/db/db.go

```diff
@@ -0,0 +1,551 @@
+	storedService.Created = service.Created
+	storedService.Size = service.Size
+	return storedService
+}
+
+// AddIOPort adds input and output ports to the database
+func (d *Db) AddIOPort(service Service) error {
+	var err error
+	for _, p := range service.Input {
+		var curInputPort IOPortTable
+		curInputPort.Path = p.Path
+		curInputPort.Name = p.Name
+		curInputPort.ID = p.ID
+		curInputPort.IsInput = true
+		curInputPort.ServiceID = string(service.ID)
+		err = d.Insert(&curInputPort)
+		if err != nil {
+			break
+		}
+	}
+	if err == nil {
+		for _, p := range service.Output {
+			var curOutputPort IOPortTable
+			curOutputPort.Path = p.Path
+			curOutputPort.Name = p.Name
+			curOutputPort.ID = p.ID
+			curOutputPort.IsInput = false
+			curOutputPort.ServiceID = string(service.ID)
+			err = d.Insert(&curOutputPort)
+			if err != nil {
+				break
+			}
+		}
+	}
+
+	return err
+}
+
+// AddService creates a new service in the database
+func (d *Db) AddService(service Service) error {
+	// Before adding a service we need to check if the service with the same name already exists.
+	// If it does, we remove it and add a new one
+	var servicesFromTable []ServiceTable
+	_, err := d.Select(&servicesFromTable, "SELECT * FROM services WHERE Name=?", service.Name)
+	if len(servicesFromTable) > 0 {
+		for _, s := range servicesFromTable {
+			err = d.RemoveService(ServiceID(s.ID))
+		}
+	}
+	if err == nil {
+		err = d.AddIOPort(service)
+		if err == nil {
+			storedService := d.MapJSON2StoredService(service)
+			err = d.Insert(&storedService)
+		}
+	}
+
+	return err
+}
+
+// RemoveService removes a service and the corresponding IOPorts from the database
+func (d *Db) RemoveService(serviceID ServiceID) error {
+	_, err := d.Exec("DELETE FROM IOPorts WHERE ServiceID=?", string(serviceID))
+	if err == nil {
+		_, err = d.Exec("DELETE FROM services WHERE ID=?", string(serviceID))
+	}
+	return err
+}
+
+// ListServices produces a list of all services ready to be converted into JSON
+func (d *Db) ListServices() ([]Service, error) {
+	var services []Service
+	var servicesFromTable []ServiceTable
+	_, err := d.Select(&servicesFromTable, "SELECT * FROM services ORDER BY ID")
+	if err == nil {
+		for _, s := range servicesFromTable {
+			var curService Service
+			curService, err = d.MapStoredService2JSON(ServiceID(s.ID), s)
+			if err == nil {
+				services = append(services, curService)
+			} else {
+				break
+			}
+		}
+
+	}
+
+	return services, err
+}
+
+// GetService returns a service ready to be converted into JSON
+func (d *Db) GetService(serviceID ServiceID) (Service, error) {
+	var service Service
+	var serviceFromTable ServiceTable
+	err := d.SelectOne(&serviceFromTable, "SELECT * FROM services WHERE ID=?", string(serviceID))
+
+	if err == nil {
+		service, err = d.MapStoredService2JSON(ServiceID(serviceFromTable.ID), serviceFromTable)
+	}
+
+	return service, err
+}
+
+// NewServiceFromImage extracts metadata and creates a valid GEF service
```

Comment:
this function and the next one should not be in this file, they have no relation to the db